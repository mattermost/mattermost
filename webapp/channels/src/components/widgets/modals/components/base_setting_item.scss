.mm-modal-generic-section-item {
    display: flex;
    flex-direction: column;
    gap: 8px;

    &__title {
        display: flex;
        align-items: center;
        padding: 0;
        margin: 0 0 8px;
        color: var(--center-channel-color);
        font-family: "Open Sans", sans-serif;
        font-size: 14px;
        font-style: normal;
        font-weight: 600;
        line-height: 20px;
    }

    &__description {
        display: flex;
        align-items: center;
        padding: 0;
        margin: 0;
        color: rgba(var(--center-channel-color-rgb), 0.64);
        font-family: 'Open Sans', sans-serif;
        font-size: 12px;
        font-style: normal;
        font-weight: 400;
        line-height: 16px;
    }

    &__error {
        display: flex;
        align-items: center;
        padding: 0;
        margin: 12px 0 0 0;
        color: var(--error-text);
        font-family: 'Open Sans', sans-serif;
        font-size: 12px;
        font-style: normal;
        font-weight: 400;
        line-height: 16px;

        svg {
            flex-shrink: 0;
            margin-right: 6px;
        }
    }

    &__content {
        display: flex;
        flex-direction: column;

        &.inline {
            flex-direction: row;
            gap: 24px;
        }
    }

    &__fieldset-checkbox-ctr {
        display: flex;
        flex-direction: row;
        align-items: center;
        margin-bottom: 5px;
        gap: 8px;
    }

    &__fieldset-checkbox {
        display: flex;
        flex-direction: row;
        align-items: center;
        margin-bottom: 0;
        cursor: pointer;
        font-size: 14px;
        gap: 8px;
        line-height: 20px;

        span {
            font-weight: 400;
        }
    }

    &__fieldset-radio {
        display: flex;
        flex-direction: column;
        gap: 12px;
    }

    &__label-radio {
        display: flex;
        width: fit-content;
        flex-direction: row;
        align-items: center;
        margin-bottom: 0;
        cursor: pointer;
        font-size: 14px;
        font-weight: 400;
        gap: 8px;
        line-height: 20px;
    }

    &__fieldset-react-select {
        max-width: 240px;
        flex-grow: 1;
    }

    &__input-checkbox {
        width: 1.6rem;
        height: 1.6rem;
    }

    input[type=checkbox] {
        display: grid;
        border: 1px solid rgba(var(--center-channel-color-rgb), 0.24);
        border-radius: 3px;
        margin: 0;
        -webkit-appearance: none;
        appearance: none;
        background-color: transparent;
        color: rgba(var(--center-channel-color-rgb), 0.24);
        cursor: pointer;
        font: inherit;
        place-content: center;
        -webkit-transition: background-color 200ms ease-out;
        -moz-transition: background-color 200ms ease-out;
        -o-transition: background-color 200ms ease-out;
        transition: background-color 200ms ease-out;
    }

    input[type="checkbox"]::before {
        width: 10px;
        height: 10px;
        background: var(--button-color);
        clip-path: polygon(0% 57%, 32% 88%, 100%  20%, 88% 8%, 32% 64%, 12% 45%);
        content: '';
        transform: scale(0);
        transform-origin: center center;
        transition: 200ms transform ease-in-out;
    }

    input[type="checkbox"]:checked::before {
        transform: scale(1);
    }

    input[type="checkbox"]:checked {
        border-color: var(--denim-button-bg);
        background: var(--denim-button-bg);
    }

    input[type=radio] {
        display: grid;
        width: 1.6rem;
        height: 1.6rem;
        border: 1px solid rgba(var(--center-channel-color-rgb), 0.24);
        border-radius: 50%;
        margin: 0;
        -webkit-appearance: none;
        appearance: none;
        background-color: transparent;
        color: rgba(var(--center-channel-color-rgb), 0.24);
        cursor: pointer;
        font: inherit;
        place-content: center;
    }

    input[type="radio"]::before {
        width: 8px;
        height: 8px;
        border-radius: 50%;
        background: var(--denim-button-bg);
        content: '';
        transform: scale(0);
        transform-origin: center center;
        transition: 200ms transform ease-in-out;
    }

    input[type="radio"]:checked::before {
        transform: scale(1);
    }

    input[type="radio"]:checked {
        border-color: var(--denim-button-bg);
    }
}
