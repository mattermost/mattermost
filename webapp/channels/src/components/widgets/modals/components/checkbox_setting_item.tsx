// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {ReactNode} from 'react';
import React from 'react';

import type {BaseSettingItemProps} from './base_setting_item';
import BaseSettingItem from './base_setting_item';

export type FieldsetCheckbox = {
    dataTestId?: string;
    name: string;
}

type Props = BaseSettingItemProps & {
    inputFieldData: FieldsetCheckbox;
    inputFieldValue: boolean;

    /**
     * The title of the checkbox input field, pass in FormattedMessage component for styling compatibility
     */
    inputFieldTitle: ReactNode;
    handleChange: (e: boolean) => void;
    className?: string;
    descriptionAboveContent?: boolean;
}

export default function CheckboxSettingItem({
    title,
    description,
    inputFieldData,
    inputFieldValue,
    inputFieldTitle,
    handleChange,
    className,
    dataTestId,
    descriptionAboveContent = false,
}: Props) {
    const content = (
        <div
            key={inputFieldData.name}
            className='mm-modal-generic-section-item__fieldset-checkbox-ctr'
        >
            <input
                className='mm-modal-generic-section-item__input-checkbox'
                data-testid={inputFieldData.dataTestId}
                type='checkbox'
                name={inputFieldData.name}
                id={inputFieldData.name.replaceAll(' ', '-')}
                checked={inputFieldValue}
                onChange={(e) => handleChange(e.target.checked)}
            />
            <label
                htmlFor={inputFieldData.name.replaceAll(' ', '-')}
                className='mm-modal-generic-section-item__fieldset-checkbox'
            >
                {inputFieldTitle}
            </label>
        </div>
    );

    return (
        <BaseSettingItem
            content={content}
            title={title}
            description={description}
            dataTestId={dataTestId}
            className={className}
            descriptionAboveContent={descriptionAboveContent}
        />
    );
}
