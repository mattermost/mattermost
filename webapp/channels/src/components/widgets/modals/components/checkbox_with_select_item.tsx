// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {ReactNode} from 'react';
import React from 'react';
import ReactSelect from 'react-select';
import type {ValueType} from 'react-select';

import type {BaseSettingItemProps} from './base_setting_item';
import BaseSettingItem from './base_setting_item';
import type {FieldsetCheckbox} from './checkbox_setting_item';
import type {FieldsetReactSelect, Option} from './react_select_item';

type Props = BaseSettingItemProps & {
    containerClassName?: string;
    descriptionAboveContent?: boolean;
    checkboxFieldTitle: ReactNode;
    checkboxFieldData: FieldsetCheckbox;
    checkboxFieldValue: boolean;
    handleCheckboxChange: (e: boolean) => void;
    selectFieldData: FieldsetReactSelect;
    selectFieldValue?: Option;
    handleSelectChange: (selected: ValueType<Option>) => void;
    isSelectDisabled?: boolean;
    selectPlaceholder?: string;
}

export default function CheckboxWithSelectSettingItem({
    title,
    description,
    containerClassName,
    descriptionAboveContent = false,
    checkboxFieldTitle,
    checkboxFieldData,
    checkboxFieldValue,
    handleCheckboxChange,
    selectFieldData,
    selectFieldValue,
    handleSelectChange,
    isSelectDisabled,
    selectPlaceholder,
}: Props) {
    const content = (
        <>
            <fieldset
                key={checkboxFieldData.name}
                className='mm-modal-generic-section-item__fieldset-checkbox-ctr'
            >
                <label className='mm-modal-generic-section-item__fieldset-checkbox'>
                    <input
                        className='mm-modal-generic-section-item__input-checkbox'
                        data-testid={checkboxFieldData.dataTestId}
                        type='checkbox'
                        name={checkboxFieldData.name}
                        checked={checkboxFieldValue}
                        onChange={(e) => handleCheckboxChange(e.target.checked)}
                    />
                    {checkboxFieldTitle}
                </label>
            </fieldset>
            <fieldset className='mm-modal-generic-section-item__fieldset-react-select'>
                <ReactSelect
                    id={selectFieldData.id}
                    inputId={selectFieldData.inputId}
                    aria-labelledby={selectFieldData.ariaLabelledby}
                    className='react-select singleSelect react-select-top'
                    classNamePrefix='react-select'
                    options={selectFieldData.options}
                    clearable={selectFieldData.clearable}
                    isClearable={selectFieldData.clearable}
                    isDisabled={isSelectDisabled}
                    isSearchable={false}
                    placeholder={selectPlaceholder}
                    onChange={(value) => handleSelectChange(value)}
                    value={selectFieldValue}
                    components={{IndicatorSeparator: NoIndicatorSeparatorComponent}}
                />
            </fieldset>
        </>
    );

    return (
        <BaseSettingItem
            title={title}
            content={content}
            isContentInline={true}
            description={description}
            className={containerClassName}
            descriptionAboveContent={descriptionAboveContent}
        />
    );
}

function NoIndicatorSeparatorComponent() {
    return null;
}

export type SelectOption = {
    value: string;
    label: ReactNode;
};
