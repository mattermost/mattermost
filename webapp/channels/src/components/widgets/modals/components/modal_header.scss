.mm-modal-header {
    display: flex;
    align-items: center;
    padding: 20px 20px 20px 32px;
    border-bottom: var(--border-light);
    background: none;
    gap: 8px;

    &__title {
        padding: 0;
        margin: 0;
        color: var(--center-channel-color);
        font-family: 'Metropolis', serif;
        font-size: 22px;
        font-style: normal;
        font-weight: 600;
        line-height: 28px;
        white-space: nowrap;
    }

    &__vertical-divider {
        width: 1px;
        height: 24px;
        background: rgba(var(--center-channel-color-rgb), 0.16);
    }

    &__subtitle {
        overflow: hidden;
        padding: 0;
        margin: 0;
        color: rgba(var(--center-channel-color-rgb), 0.56);
        font-family: inherit;
        font-size: 12px;
        font-style: normal;
        font-weight: 400;
        line-height: 20px;
        text-overflow: ellipsis;
        white-space: nowrap;
    }

    &__ctr {
        display: flex;
        flex: 1 1 auto;
        align-items: center;
        justify-content: flex-end;
    }

    @media screen and (max-width: 768px) {
        padding: 18px 24px;

        &__title {
            font-size: 20px;
        }

        &__subtitle {
            display: none;
        }

        &__vertical-divider {
            display: none;
        }
    }
}
