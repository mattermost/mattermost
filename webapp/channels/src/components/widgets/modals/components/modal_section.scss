.mm-modal-generic-section {
    display: flex;
    flex-direction: column;

    .modalSectionHeader {
        display: flex;
        flex-direction: column;
        gap: 8px;
        margin-block-end: 24px;

        .modalSectionTitle {
            display: flex;
            align-items: center;
            padding: 0;
            margin: 0;
            color: var(--center-channel-color);
            font-family: "Metropolis", sans-serif;
            font-size: 16px;
            font-style: normal;
            font-weight: 600;
            gap: 8px;
            line-height: 24px;
        }

        .modalSectionDescription {
            display: flex;
            align-items: center;
            padding: 0;
            margin: 0;
            color: rgba(var(--center-channel-color-rgb), 0.64);
            font-family: "Open Sans", sans-serif;
            font-size: 12px;
            font-style: normal;
            font-weight: 400;
            line-height: 16px;
        }
    }

    .modalSectionContent {
        display: flex;
        flex-direction: column;
        gap: 24px;
    }
}
