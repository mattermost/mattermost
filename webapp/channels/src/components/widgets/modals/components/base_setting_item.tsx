// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import classNames from 'classnames';
import type {PrimitiveType, FormatXMLElementFn} from 'intl-messageformat';
import React from 'react';
import type {MessageDescriptor} from 'react-intl';
import {useIntl} from 'react-intl';

import {AlertCircleOutlineIcon} from '@mattermost/compass-icons/components';

import './base_setting_item.scss';

type ExtendedMessageDescriptor = MessageDescriptor & {
    values?: Record<string, PrimitiveType | FormatXMLElementFn<string, string>>;
};

export type BaseSettingItemProps = {
    title?: string;
    description?: string;
    error?: ExtendedMessageDescriptor;
    dataTestId?: string;
};

type Props = BaseSettingItemProps & {
    content: JSX.Element;
    isContentInline?: boolean;
    className?: string;
    descriptionAboveContent?: boolean;
}

function BaseSettingItem({title, description, content, className, error, descriptionAboveContent = false, isContentInline = false, dataTestId}: Props): JSX.Element {
    const {formatMessage} = useIntl();

    const titleComponent = title && (
        <h4
            data-testid='mm-modal-generic-section-item__title'
            className='mm-modal-generic-section-item__title'
        >
            {title}
        </h4>
    );

    const descriptionComponent = description && (
        <p
            data-testid='mm-modal-generic-section-item__description'
            className='mm-modal-generic-section-item__description'
        >
            {description}
        </p>
    );

    const Error = error && (
        <div
            data-testid='mm-modal-generic-section-item__error'
            className='mm-modal-generic-section-item__error'
        >
            <AlertCircleOutlineIcon/>
            {formatMessage({id: error.id, defaultMessage: error.defaultMessage}, error.values)}
        </div>
    );

    return (
        <div
            data-testid={dataTestId}
            className={classNames('mm-modal-generic-section-item', className)}
        >
            {titleComponent}
            {descriptionAboveContent ? descriptionComponent : undefined}
            <div
                data-testid='mm-modal-generic-section-item__content'
                className={classNames('mm-modal-generic-section-item__content', {
                    inline: isContentInline,
                })}
            >
                {content}
            </div>
            {descriptionAboveContent ? undefined : descriptionComponent}
            {Error}
        </div>
    );
}

export default BaseSettingItem;

