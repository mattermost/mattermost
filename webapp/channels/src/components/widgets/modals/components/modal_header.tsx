// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {useIntl} from 'react-intl';
import './modal_header.scss';

type Props = {
    id: string;
    title: string;
    subtitle: string;
    handleClose?: (e: React.MouseEvent) => void;
}

function ModalHeader({id, title, subtitle, handleClose}: Props) {
    const intl = useIntl();
    return (
        <header className='mm-modal-header'>
            <h1
                id={`mm-modal-header-${id}`}
                className='mm-modal-header__title'
            >
                {title}
            </h1>
            <div className='mm-modal-header__vertical-divider'/>
            <p className='mm-modal-header__subtitle'>{subtitle}</p>
            <div
                className='mm-modal-header__ctr'
                onClick={handleClose}
            >
                <button
                    className='btn btn-icon'
                    aria-label={intl.formatMessage({id: 'modal.header_close', defaultMessage: 'Close'})}
                >
                    <i className='icon icon-close'/>
                </button>
            </div>
        </header>
    );
}
export default ModalHeader;
