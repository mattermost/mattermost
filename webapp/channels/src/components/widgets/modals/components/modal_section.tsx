// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import type {ReactNode} from 'react';

import './modal_section.scss';

type Props = {
    title?: ReactNode;
    description?: ReactNode;
    content: JSX.Element;
    titleSuffix?: JSX.Element;
};

function ModalSection({
    title,
    description,
    content,
    titleSuffix,
}: Props) {
    const titleComponent = title && (
        <h4 className='modalSectionTitle'>
            {title}
            {titleSuffix}
        </h4>
    );

    const descriptionComponent = description && (
        <p className='modalSectionDescription'>
            {description}
        </p>
    );

    return (
        <section className='mm-modal-generic-section'>
            {(title || description) && (
                <div className='modalSectionHeader'>
                    {titleComponent}
                    {descriptionComponent}
                </div>
            )}
            <div className='modalSectionContent'>
                {content}
            </div>
        </section>
    );
}

export default ModalSection;
