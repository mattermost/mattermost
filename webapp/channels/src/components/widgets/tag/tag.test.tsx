// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {shallow} from 'enzyme';
import React from 'react';

import {AlertCircleOutlineIcon} from '@mattermost/compass-icons/components';

import Tag from './tag';

const classNameProp = 'Tag Tag--xs test';

describe('components/widgets/tag/Tag', () => {
    test('should match the snapshot on show', () => {
        const wrapper = shallow(
            <Tag
                className={'test'}
                text={'Test text'}
            />,
        );
        expect(wrapper.props()).toEqual(expect.objectContaining({className: classNameProp}));
        expect(wrapper.text()).toEqual('Test text');
        expect(wrapper).toMatchSnapshot();
    });

    test('should match the snapshot with icon', () => {
        const wrapper = shallow(
            <Tag
                className={'test'}
                text={'Test text'}
                icon={'alert-circle-outline'}
            />,
        );
        expect(wrapper.props()).toEqual(expect.objectContaining({className: classNameProp}));
        expect(wrapper.find(AlertCircleOutlineIcon).exists()).toEqual(true);
        expect(wrapper.text()).toContain('Test text');
        expect(wrapper).toMatchSnapshot();
    });

    test('should match the snapshot with uppercase prop', () => {
        const wrapper = shallow(
            <Tag
                className={'test'}
                text={'Test text'}
                uppercase={true}
            />,
        );
        expect(wrapper.props()).toEqual(expect.objectContaining({className: classNameProp, uppercase: true}));
        expect(wrapper.text()).toEqual('Test text');
        expect(wrapper).toMatchSnapshot();
    });

    test('should match the snapshot with size "sm"', () => {
        const wrapper = shallow(
            <Tag
                className={'test'}
                text={'Test text'}
                size={'sm'}
            />,
        );
        expect(wrapper.props()).toEqual(expect.objectContaining({className: 'Tag Tag--sm test'}));
        expect(wrapper.text()).toEqual('Test text');
        expect(wrapper).toMatchSnapshot();
    });

    test('should match the snapshot with "success" variant', () => {
        const wrapper = shallow(
            <Tag
                className={'test'}
                text={'Test text'}
                variant={'success'}
            />,
        );
        expect(wrapper.props()).toEqual(expect.objectContaining({className: 'Tag Tag--success Tag--xs test'}));
        expect(wrapper.text()).toEqual('Test text');
        expect(wrapper).toMatchSnapshot();
    });

    test('should transform into a button if onClick provided', () => {
        const click = jest.fn();
        const wrapper = shallow(
            <Tag
                className={'test'}
                text={'Test text'}
                onClick={click}
            />,
        );
        expect(wrapper.props()).toEqual(expect.objectContaining({className: classNameProp, onClick: click}));
        expect(wrapper.text()).toEqual('Test text');
        wrapper.simulate('click');
        expect(click).toHaveBeenCalled();
        expect(wrapper).toMatchSnapshot();
    });
});
