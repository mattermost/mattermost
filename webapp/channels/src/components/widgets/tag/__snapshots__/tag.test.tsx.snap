// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`components/widgets/tag/Tag should match the snapshot on show 1`] = `
<TagWrapper
  as="div"
  className="Tag Tag--xs test"
  uppercase={false}
>
  <TagText>
    Test text
  </TagText>
</TagWrapper>
`;

exports[`components/widgets/tag/Tag should match the snapshot with "success" variant 1`] = `
<TagWrapper
  as="div"
  className="Tag Tag--success Tag--xs test"
  uppercase={false}
>
  <TagText>
    Test text
  </TagText>
</TagWrapper>
`;

exports[`components/widgets/tag/Tag should match the snapshot with icon 1`] = `
<TagWrapper
  as="div"
  className="Tag Tag--xs test"
  uppercase={false}
>
  <AlertCircleOutlineIcon
    size={10}
  />
  <TagText>
    Test text
  </TagText>
</TagWrapper>
`;

exports[`components/widgets/tag/Tag should match the snapshot with size "sm" 1`] = `
<TagWrapper
  as="div"
  className="Tag Tag--sm test"
  uppercase={false}
>
  <TagText>
    Test text
  </TagText>
</TagWrapper>
`;

exports[`components/widgets/tag/Tag should match the snapshot with uppercase prop 1`] = `
<TagWrapper
  as="div"
  className="Tag Tag--xs test"
  uppercase={true}
>
  <TagText>
    Test text
  </TagText>
</TagWrapper>
`;

exports[`components/widgets/tag/Tag should transform into a button if onClick provided 1`] = `
<TagWrapper
  as="button"
  className="Tag Tag--xs test"
  onClick={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  uppercase={false}
>
  <TagText>
    Test text
  </TagText>
</TagWrapper>
`;
